torch/utils/data/graph.py

```diff
@@ -40,17 +40,17 @@ def reduce_hook(obj):
             captured_connections.append(obj)
             return stub_unpickler, ()
 
-    # TODO(VitalyFedyunin):  Better do it as `with` context for safety
-    IterDataPipe.set_reduce_ex_hook(reduce_hook)
-    if exclude_primitive:
-        IterDataPipe.set_getstate_hook(getstate_hook)
     try:
```

Comment:
I am under the impression that using `try ... finally ...` here will provide more safety. The `reduce` and `getstate` hooks will be re-set to `None` even if an exception is thrown.

Does that seem better? Or do we have to use a `with` statement?